docs/api/main-config-indexers.md

```diff
@@ -0,0 +1,352 @@
+
+Transpiling the custom source format to CSF is beyond the scope of this documentation. This transpilation is often done at the builder level ([Vite](../builders/vite.md) and/or [Webpack](../builders/webpack.md)), and we recommend using [unplugin](https://github.com/unjs/unplugin) to create plugins for multiple builders.
+
+The general architecture looks something like this:
+
+<!-- TODO: If diagram is useful, make a real one -->
+
+![Architecture diagram showing how custom indexers are used to generate stories from source files](./main-config-indexers-architecture.png)
+
+First, here's an example of a source file:
+
+```ts
+// ./Button.variants.ts
+
+import { variantsFromComponent, createStoryFromVariant } from '../utils';
+import { Button } from './Button';
+
+/**
+ * Returns raw strings representing stories via component props, eg.
+ * 'export const PrimaryVariant = {
+ *    args: {
+ *      primary: true
+ *    },
+ *  };'
+ */
+export const generateStories = () => {
+  const variants = variantsFromComponent(Button);
+  return variants.map((variant) => createStoryFromVariant(variant));
+};
+```
+
+The builder plugin would then:
+
+1. Receive and read the source file
+2. Import the exported `generateStories` function
+3. Run the function to generate the stories
+4. Write the stories to a CSF file
+
+That resulting CSF file would then be indexed by Storybook. It would something look like this:
+
+```js
+// virtual:./Button.variants
+
+import { Button } from './Button';
+
+export default {
+  component: Button,
+};
+
+export const Primary = {
+  args: {
+    primary: true,
+  },
+};
+```
+
+### Examples
+
+Some example usages of custom indexers include:
+
+<details open>
+
+<summary>Generating stories dynamically from fixture data or API endpoints</summary>
+
+This indexer generates stories for components based on JSON fixture data. It looks for `*.stories.json` files in the project, adds them to the index and separately converts their content to CSF.
+
+<!-- prettier-ignore-start -->
+
+<CodeSnippets
+  paths={[
+    'common/main-config-indexers-jsonstories.js.mdx',
+    'common/main-config-indexers-jsonstories.ts.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+An example input JSON file could look like this:
+
+```json
+{
+  "Button": {
+    "componentPath": "./button/Button.jsx",
+    "stories": {
+      "Primary": {
+        "args": {
+          "primary": true
+        },
+      "Secondary": {
+        "args": {
+          "primary": false
+        }
+      }
+    }
+  },
+  "Dialog": {
+    "componentPath": "./dialog/Dialog.jsx",
+    "stories": {
+      "Closed": {},
+      "Open": {
+        "args": {
+          "isOpen": true
+        }
+      },
+    }
+  }
+}
+```
+
+A builder plugin will then need to transform the JSON file into a regular CSF file. This transformation could be done with a Vite plugin similar to this:
+
+```ts
```

Comment:
@JReinhold can you add a comment mentioning the file location/name so that we can provide some context on it? Also, just a quick question to help @kylegach handle the snippets, can a user write this as Javascript code, or should it only be TS-specific?

Reply:
JavaScript is totally acceptable.

Reply:
> handle the snippets, can a user write this as Javascript code, or should it only be TS-specific?

I opted to _not_ provide JS/TS snippets for this, instead just leaving it as a single code block. The code here is for a builder plugin that doesn't use any Storybook API. We're really only including it as a courtesy example for those brave enough to attempt a custom indexer.

Reply:
Thanks for the clarification. Appreciate it 🙏 ! And we may be surprised as there are some brave souls out there willing to give it a go and start building a custom indexer.